File: backend/app/api/patients.py
"""
API routes para gestão de pacientes
"""

from flask import Blueprint, request, jsonify, current_app
from flask_jwt_extended import jwt_required, get_jwt_identity
from sqlalchemy import or_, and_
from sqlalchemy.orm import joinedload
from datetime import datetime, date
from typing import Dict, Any, List, Optional

from app import db
from app.models.user import User, UserRole
from app.models.patient import Patient, MedicalRecord, Evolution, EmergencyContact
from app.auth.utils import roles_required
from app.utils.validation import validate_cpf, validate_phone, validate_email

patients_bp = Blueprint('patients', __name__)


@patients_bp.route('/', methods=['POST'])
@roles_required(UserRole.ADMIN, UserRole.FISIOTERAPEUTA, UserRole.ESTAGIARIO)
def create_patient():
    """Criar novo paciente"""
    try:
        data = request.get_json()
        
        # Validações básicas
        if not data.get('nome_completo'):
            return jsonify({'error': 'Nome completo é obrigatório'}), 400
        
        # Validar CPF se fornecido
        if data.get('cpf') and not validate_cpf(data['cpf']):
            return jsonify({'error': 'CPF inválido'}), 400
        
        # Validar telefone se fornecido
        if data.get('telefone') and not validate_phone(data['telefone']):
            return jsonify({'error': 'Telefone inválido'}), 400
        
        # Validar email se fornecido
        if data.get('email') and not validate_email(data['email']):
            return jsonify({'error': 'Email inválido'}), 400
        
        # Verificar se CPF já existe (se fornecido)
        if data.get('cpf'):
            existing_patient = Patient.query.filter(
                Patient.cpf_encrypted.isnot(None)
            ).all()
            
            for patient in existing_patient:
                if patient.cpf == data['cpf']:
                    return jsonify({'error': 'Paciente com este CPF já existe'}), 409
        
        # Criar paciente
        patient = Patient(
            nome_completo=data['nome_completo'],
            nome_social=data.get('nome_social'),
            data_nascimento=datetime.strptime(data['data_nascimento'], '%Y-%m-%d').date() if data.get('data_nascimento') else None,
            genero=data.get('genero'),
            tipo_sanguineo=data.get('tipo_sanguineo'),
            estado_civil=data.get('estado_civil'),
            escolaridade=data.get('escolaridade'),
            profissao=data.get('profissao'),
            email=data.get('email'),
            endereco=data.get('endereco'),
            observacoes=data.get('observacoes'),
            consentimento_dados=data.get('consentimento_dados', False),
            consentimento_imagem=data.get('consentimento_imagem', False),
            data_consentimento=datetime.utcnow() if data.get('consentimento_dados') else None,
        )
        
        # Dados criptografados
        if data.get('cpf'):
            patient.cpf = data['cpf']
        if data.get('rg'):
            patient.rg = data['rg']
        if data.get('telefone'):
            patient.telefone = data['telefone']
        if data.get('telefone_alternativo'):
            patient.telefone_alternativo = data['telefone_alternativo']
        
        db.session.add(patient)
        db.session.commit()
        
        # Adicionar contatos de emergência se fornecidos
        if data.get('emergency_contacts'):
            for contact_data in data['emergency_contacts']:
                contact = EmergencyContact(
                    patient_id=patient.id,
                    nome=contact_data['nome'],
                    tipo_contato=contact_data['tipo_contato'],
                    parentesco=contact_data.get('parentesco'),
                    email=contact_data.get('email'),
                    endereco=contact_data.get('endereco'),
                    observacoes=contact_data.get('observacoes'),
                )
                
                if contact_data.get('telefone'):
                    contact.telefone = contact_data['telefone']
                
                db.session.add(contact)
        
        db.session.commit()
        
        return jsonify({
            'message': 'Paciente criado com sucesso',
            'patient': patient.to_dict()
        }), 201
        
    except Exception as e:
        db.session.rollback()
        current_app.logger.error(f"Erro ao criar paciente: {e}")
        return jsonify({'error': 'Erro interno do servidor'}), 500


@patients_bp.route('/', methods=['GET'])
@roles_required(UserRole.ADMIN, UserRole.FISIOTERAPEUTA, UserRole.ESTAGIARIO)
def list_patients():
    """Listar pacientes com paginação e filtros"""
    try:
        # Parâmetros de paginação
        page = request.args.get('page', 1, type=int)
        per_page = min(request.args.get('per_page', 20, type=int), 100)
        
        # Filtros
        search = request.args.get('search', '').strip()
        is_active = request.args.get('is_active', type=bool)
        
        # Query base
        query = Patient.query
        
        # Aplicar filtros
        if search:
            query = query.filter(
                or_(
                    Patient.nome_completo.ilike(f'%{search}%'),
                    Patient.nome_social.ilike(f'%{search}%'),
                    Patient.email.ilike(f'%{search}%'),
                )
            )
        
        if is_active is not None:
            query = query.filter(Patient.is_active == is_active)
        
        # Ordenação
        query = query.order_by(Patient.nome_completo.asc())
        
        # Paginação
        pagination = query.paginate(
            page=page,
            per_page=per_page,
            error_out=False
        )
        
        patients = [patient.to_dict() for patient in pagination.items]
        
        return jsonify({
            'patients': patients,
            'pagination': {
                'page': page,
                'pages': pagination.pages,
                'per_page': per_page,
                'total': pagination.total,
                'has_prev': pagination.has_prev,
                'has_next': pagination.has_next,
            }
        }), 200
        
    except Exception as e:
        current_app.logger.error(f"Erro ao listar pacientes: {e}")
        return jsonify({'error': 'Erro interno do servidor'}), 500


@patients_bp.route('/<patient_id>', methods=['GET'])
@roles_required(UserRole.ADMIN, UserRole.FISIOTERAPEUTA, UserRole.ESTAGIARIO, UserRole.PACIENTE)
def get_patient(patient_id: str):
    """Obter detalhes de um paciente"""
    try:
        current_user_id = get_jwt_identity()
        current_user = User.query.get(current_user_id)
        
        # Verificar se usuário pode acessar este paciente
        patient = Patient.query.get_or_404(patient_id)
        
        # Pacientes só podem ver seus próprios dados
        if current_user.role == UserRole.PACIENTE:
            if not patient.user_id or patient.user_id != current_user_id:
                return jsonify({'error': 'Acesso negado'}), 403
        
        # Incluir dados sensíveis apenas para profissionais
        include_sensitive = current_user.role in [UserRole.ADMIN, UserRole.FISIOTERAPEUTA, UserRole.ESTAGIARIO]
        
        # Carregar relacionamentos
        patient_data = patient.to_dict(include_sensitive=include_sensitive)
        
        # Adicionar contatos de emergência
        emergency_contacts = EmergencyContact.query.filter_by(patient_id=patient_id).all()
        patient_data['emergency_contacts'] = [
            contact.to_dict(include_sensitive=include_sensitive) 
            for contact in emergency_contacts
        ]
        
        # Adicionar histórico médico (apenas para profissionais)
        if include_sensitive:
            medical_records = MedicalRecord.query.filter_by(patient_id=patient_id).order_by(
                MedicalRecord.data_avaliacao.desc()
            ).all()
            patient_data['medical_records'] = [record.to_dict() for record in medical_records]
        
        return jsonify(patient_data), 200
        
    except Exception as e:
        current_app.logger.error(f"Erro ao obter paciente: {e}")
        return jsonify({'error': 'Erro interno do servidor'}), 500


@patients_bp.route('/<patient_id>', methods=['PUT'])
@roles_required(UserRole.ADMIN, UserRole.FISIOTERAPEUTA, UserRole.ESTAGIARIO)
def update_patient(patient_id: str):
    """Atualizar dados do paciente"""
    try:
        patient = Patient.query.get_or_404(patient_id)
        data = request.get_json()
        
        # Validações
        if data.get('cpf') and not validate_cpf(data['cpf']):
            return jsonify({'error': 'CPF inválido'}), 400
        
        if data.get('telefone') and not validate_phone(data['telefone']):
            return jsonify({'error': 'Telefone inválido'}), 400
        
        if data.get('email') and not validate_email(data['email']):
            return jsonify({'error': 'Email inválido'}), 400
        
        # Verificar CPF duplicado se mudou
        if data.get('cpf') and data['cpf'] != patient.cpf:
            existing_patient = Patient.query.filter(
                and_(
                    Patient.id != patient_id,
                    Patient.cpf_encrypted.isnot(None)
                )
            ).all()
            
            for existing in existing_patient:
                if existing.cpf == data['cpf']:
                    return jsonify({'error': 'Paciente com este CPF já existe'}), 409
        
        # Atualizar campos
        updateable_fields = [
            'nome_completo', 'nome_social', 'genero', 'tipo_sanguineo',
            'estado_civil', 'escolaridade', 'profissao', 'email',
            'endereco', 'observacoes', 'is_active'
        ]
        
        for field in updateable_fields:
            if field in data:
                setattr(patient, field, data[field])
        
        # Campos com tratamento especial
        if 'data_nascimento' in data and data['data_nascimento']:
            patient.data_nascimento = datetime.strptime(data['data_nascimento'], '%Y-%m-%d').date()
        
        # Campos criptografados
        if 'cpf' in data:
            patient.cpf = data['cpf']
        if 'rg' in data:
            patient.rg = data['rg']
        if 'telefone' in data:
            patient.telefone = data['telefone']
        if 'telefone_alternativo' in data:
            patient.telefone_alternativo = data['telefone_alternativo']
        
        # Atualizar consentimentos
        if 'consentimento_dados' in data:
            patient.consentimento_dados = data['consentimento_dados']
            if data['consentimento_dados'] and not patient.data_consentimento:
                patient.data_consentimento = datetime.utcnow()
        
        if 'consentimento_imagem' in data:
            patient.consentimento_imagem = data['consentimento_imagem']
        
        patient.updated_at = datetime.utcnow()
        db.session.commit()
        
        return jsonify({
            'message': 'Paciente atualizado com sucesso',
            'patient': patient.to_dict()
        }), 200
        
    except Exception as e:
        db.session.rollback()
        current_app.logger.error(f"Erro ao atualizar paciente: {e}")
        return jsonify({'error': 'Erro interno do servidor'}), 500


@patients_bp.route('/<patient_id>', methods=['DELETE'])
@roles_required(UserRole.ADMIN)
def delete_patient(patient_id: str):
    """Excluir paciente (soft delete)"""
    try:
        patient = Patient.query.get_or_404(patient_id)
        
        # Soft delete - apenas desativa o paciente
        patient.is_active = False
        patient.updated_at = datetime.utcnow()
        
        db.session.commit()
        
        return jsonify({'message': 'Paciente desativado com sucesso'}), 200
        
    except Exception as e:
        db.session.rollback()
        current_app.logger.error(f"Erro ao excluir paciente: {e}")
        return jsonify({'error': 'Erro interno do servidor'}), 500


@patients_bp.route('/<patient_id>/emergency-contacts', methods=['POST'])
@roles_required(UserRole.ADMIN, UserRole.FISIOTERAPEUTA, UserRole.ESTAGIARIO)
def add_emergency_contact(patient_id: str):
    """Adicionar contato de emergência"""
    try:
        patient = Patient.query.get_or_404(patient_id)
        data = request.get_json()
        
        if not data.get('nome'):
            return jsonify({'error': 'Nome é obrigatório'}), 400
        
        if not data.get('tipo_contato'):
            return jsonify({'error': 'Tipo de contato é obrigatório'}), 400
        
        contact = EmergencyContact(
            patient_id=patient_id,
            nome=data['nome'],
            tipo_contato=data['tipo_contato'],
            parentesco=data.get('parentesco'),
            email=data.get('email'),
            endereco=data.get('endereco'),
            observacoes=data.get('observacoes'),
        )
        
        if data.get('telefone'):
            contact.telefone = data['telefone']
        
        db.session.add(contact)
        db.session.commit()
        
        return jsonify({
            'message': 'Contato de emergência adicionado com sucesso',
            'contact': contact.to_dict(include_sensitive=True)
        }), 201
        
    except Exception as e:
        db.session.rollback()
        current_app.logger.error(f"Erro ao adicionar contato: {e}")
        return jsonify({'error': 'Erro interno do servidor'}), 500


@patients_bp.route('/emergency-contacts/<contact_id>', methods=['PUT'])
@roles_required(UserRole.ADMIN, UserRole.FISIOTERAPEUTA, UserRole.ESTAGIARIO)
def update_emergency_contact(contact_id: str):
    """Atualizar contato de emergência"""
    try:
        contact = EmergencyContact.query.get_or_404(contact_id)
        data = request.get_json()
        
        # Atualizar campos
        updateable_fields = [
            'nome', 'tipo_contato', 'parentesco', 'email', 
            'endereco', 'observacoes'
        ]
        
        for field in updateable_fields:
            if field in data:
                setattr(contact, field, data[field])
        
        if 'telefone' in data:
            contact.telefone = data['telefone']
        
        db.session.commit()
        
        return jsonify({
            'message': 'Contato atualizado com sucesso',
            'contact': contact.to_dict(include_sensitive=True)
        }), 200
        
    except Exception as e:
        db.session.rollback()
        current_app.logger.error(f"Erro ao atualizar contato: {e}")
        return jsonify({'error': 'Erro interno do servidor'}), 500


@patients_bp.route('/emergency-contacts/<contact_id>', methods=['DELETE'])
@roles_required(UserRole.ADMIN, UserRole.FISIOTERAPEUTA, UserRole.ESTAGIARIO)
def delete_emergency_contact(contact_id: str):
    """Excluir contato de emergência"""
    try:
        contact = EmergencyContact.query.get_or_404(contact_id)
        
        db.session.delete(contact)
        db.session.commit()
        
        return jsonify({'message': 'Contato removido com sucesso'}), 200
        
    except Exception as e:
        db.session.rollback()
        current_app.logger.error(f"Erro ao remover contato: {e}")
        return jsonify({'error': 'Erro interno do servidor'}), 500

File: frontend/src/services/patients.ts
'use client';

import { apiService } from './api';

// Types for Patient Management
export interface Patient {
  id: string;
  user_id?: string;
  nome_completo: string;
  nome_social?: string;
  cpf_masked?: string;
  telefone_masked?: string;
  data_nascimento?: string;
  genero?: 'MASCULINO' | 'FEMININO' | 'OUTRO' | 'NAO_INFORMADO';
  tipo_sanguineo?: string;
  estado_civil?: string;
  escolaridade?: string;
  profissao?: string;
  email?: string;
  endereco?: any;
  is_active: boolean;
  observacoes?: string;
  consentimento_dados: boolean;
  consentimento_imagem: boolean;
  data_consentimento?: string;
  age?: number;
  created_at: string;
  updated_at?: string;
  emergency_contacts?: EmergencyContact[];
  medical_records?: MedicalRecord[];
}

export interface EmergencyContact {
  id: string;
  nome: string;
  tipo_contato: 'FAMILIAR' | 'AMIGO' | 'COLEGA' | 'CUIDADOR' | 'OUTRO';
  parentesco?: string;
  telefone_masked?: string;
  email?: string;
  endereco?: any;
  observacoes?: string;
  created_at: string;
}

export interface MedicalRecord {
  id: string;
  patient_id: string;
  created_by: string;
  queixa_principal?: string;
  historia_doenca_atual?: string;
  historia_patologica_pregressa?: string;
  historia_familiar?: string;
  medicamentos_uso?: string;
  alergias?: string;
  inspecao?: string;
  palpacao?: string;
  amplitude_movimento?: any;
  forca_muscular?: any;
  testes_especiais?: any;
  escalas_funcionais?: any;
  objetivos_tratamento?: string;
  diagnostico_fisioterapeutico?: string;
  cid10?: string;
  plano_tratamento?: string;
  prognostico?: string;
  data_avaliacao: string;
  created_at: string;
  updated_at?: string;
  evolutions?: Evolution[];
}

export interface Evolution {
  id: string;
  medical_record_id: string;
  created_by: string;
  subjetivo?: string;
  objetivo?: string;
  avaliacao?: string;
  plano?: string;
  data_atendimento: string;
  duracao_minutos?: number;
  tecnicas_utilizadas?: any;
  exercicios_realizados?: any;
  escala_dor?: number;
  observacoes_clinicas?: string;
  created_at: string;
  updated_at?: string;
}

export interface CreatePatientData {
  nome_completo: string;
  nome_social?: string;
  cpf?: string;
  rg?: string;
  data_nascimento?: string;
  genero?: string;
  tipo_sanguineo?: string;
  estado_civil?: string;
  escolaridade?: string;
  profissao?: string;
  telefone?: string;
  telefone_alternativo?: string;
  email?: string;
  endereco?: any;
  observacoes?: string;
  consentimento_dados: boolean;
  consentimento_imagem?: boolean;
  emergency_contacts?: CreateEmergencyContactData[];
}

export interface CreateEmergencyContactData {
  nome: string;
  tipo_contato: string;
  parentesco?: string;
  telefone?: string;
  email?: string;
  endereco?: any;
  observacoes?: string;
}

export interface PaginatedPatientsResponse {
  patients: Patient[];
  pagination: {
    page: number;
    pages: number;
    per_page: number;
    total: number;
    has_prev: boolean;
    has_next: boolean;
  };
}

export interface PatientStats {
  patient_name: string;
  total_records: number;
  total_evolutions: number;
  latest_evaluation?: string;
  pain_evolution: Array<{
    date: string;
    pain_level: number;
  }>;
  treatment_duration_days: number;
  average_session_duration: number;
}

class PatientService {
  // Patient CRUD
  async createPatient(data: CreatePatientData): Promise<{ message: string; patient: Patient }> {
    return await apiService.request({
      method: 'POST',
      url: '/api/v1/patients/',
      data,
    });
  }

  async getPatients(params?: {
    page?: number;
    per_page?: number;
    search?: string;
    is_active?: boolean;
  }): Promise<PaginatedPatientsResponse> {
    return await apiService.request({
      method: 'GET',
      url: '/api/v1/patients/',
      params,
    });
  }

  async getPatient(patientId: string): Promise<Patient> {
    return await apiService.request({
      method: 'GET',
      url: `/api/v1/patients/${patientId}`,
    });
  }

  async updatePatient(patientId: string, data: Partial<CreatePatientData>): Promise<{ message: string; patient: Patient }> {
    return await apiService.request({
      method: 'PUT',
      url: `/api/v1/patients/${patientId}`,
      data,
    });
  }

  async deletePatient(patientId: string): Promise<{ message: string }> {
    return await apiService.request({
      method: 'DELETE',
      url: `/api/v1/patients/${patientId}`,
    });
  }

  // Emergency Contacts
  async addEmergencyContact(patientId: string, data: CreateEmergencyContactData): Promise<{ message: string; contact: EmergencyContact }> {
    return await apiService.request({
      method: 'POST',
      url: `/api/v1/patients/${patientId}/emergency-contacts`,
      data,
    });
  }

  async updateEmergencyContact(contactId: string, data: Partial<CreateEmergencyContactData>): Promise<{ message: string; contact: EmergencyContact }> {
    return await apiService.request({
      method: 'PUT',
      url: `/api/v1/patients/emergency-contacts/${contactId}`,
      data,
    });
  }

  async deleteEmergencyContact(contactId: string): Promise<{ message: string }> {
    return await apiService.request({
      method: 'DELETE',
      url: `/api/v1/patients/emergency-contacts/${contactId}`,
    });
  }

  // Medical Records
  async createMedicalRecord(data: any): Promise<{ message: string; medical_record: MedicalRecord }> {
    return await apiService.request({
      method: 'POST',
      url: '/api/v1/medical-records/',
      data,
    });
  }

  async getMedicalRecord(recordId: string): Promise<MedicalRecord> {
    return await apiService.request({
      method: 'GET',
      url: `/api/v1/medical-records/${recordId}`,
    });
  }

  async updateMedicalRecord(recordId: string, data: any): Promise<{ message: string; medical_record: MedicalRecord }> {
    return await apiService.request({
      method: 'PUT',
      url: `/api/v1/medical-records/${recordId}`,
      data,
    });
  }

  async getPatientMedicalRecords(patientId: string): Promise<{ medical_records: MedicalRecord[]; patient_name: string }> {
    return await apiService.request({
      method: 'GET',
      url: `/api/v1/medical-records/patient/${patientId}`,
    });
  }

  // Evolutions
  async createEvolution(recordId: string, data: any): Promise<{ message: string; evolution: Evolution }> {
    return await apiService.request({
      method: 'POST',
      url: `/api/v1/medical-records/${recordId}/evolutions`,
      data,
    });
  }

  async updateEvolution(evolutionId: string, data: any): Promise<{ message: string; evolution: Evolution }> {
    return await apiService.request({
      method: 'PUT',
      url: `/api/v1/medical-records/evolutions/${evolutionId}`,
      data,
    });
  }

  async deleteEvolution(evolutionId: string): Promise<{ message: string }> {
    return await apiService.request({
      method: 'DELETE',
      url: `/api/v1/medical-records/evolutions/${evolutionId}`,
    });
  }

  // Stats
  async getPatientStats(patientId: string): Promise<PatientStats> {
    return await apiService.request({
      method: 'GET',
      url: `/api/v1/medical-records/stats/${patientId}`,
    });
  }
}

export const patientService = new PatientService();
export default patientService;